File: Reaktoro/Thermodynamics/Surface/ActivityModelIonExchange.cpp
// Reaktoro is a unified framework for modeling chemically reactive systems.
//
// Copyright © 2014-2021 Allan Leal
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this library. If not, see <http://www.gnu.org/licenses/>.

#include "ActivityModelIonExchange.hpp"

// Reaktoro includes
#include <Reaktoro/Singletons/Elements.hpp>
#include <Reaktoro/Thermodynamics/Aqueous/AqueousProps.hpp>
#include <Reaktoro/Thermodynamics/Aqueous/AqueousMixture.hpp>
#include <Reaktoro/Thermodynamics/Surface/IonExchangeSurface.hpp>
#include <Reaktoro/Extensions/Phreeqc/PhreeqcLegacy.hpp>

namespace Reaktoro {

using std::sqrt;
using std::log;

namespace detail {

/// Return the IonExchangeActivityModel object based on the Gaines--Thomas model.
auto activityModelIonExchangeGainesThomas(const SpeciesList& species) -> ActivityModel
{
    // Create the ion exchange surface
    IonExchangeSurface surface(species);

    // The number of all species and ion exchange species in the current exchange phase only
    const auto num_species = species.size();

    // The numbers of exchanger's equivalents for exchange species
    ArrayXd ze = surface.ze();

    // Define the activity model function of the ion exchange phase
    ActivityModel fn = [=](ActivityPropsRef props, ActivityArgs args) mutable
    {
        // The arguments for the activity model evaluation
        const auto& [T, P, x] = args;

        // Auxiliary references
        auto& ln_g = props.ln_g;
        auto& ln_a = props.ln_a;

        // Calculate ln of activities of ion exchange species as the ln of equivalence fractions
        ln_a = (x*ze/(x*ze).sum()).log();

        // Initialized the ln of activity coefficients of the ion exchange species on the surface
        ln_g = ArrayXr::Zero(num_species);

        // Calculate Davies and Debye--Huckel parameters only if the AqueousPhase has been already evaluated
        if (props.extra["AqueousMixtureState"].has_value())
        {
            // Export aqueous mixture state via `extra` data member
            const auto& aqstate = std::any_cast<AqueousMixtureState>(props.extra["AqueousMixtureState"]);

            // Auxiliary constant references properties
            const auto& I = aqstate.Is;            // the stoichiometric ionic strength
            const auto& rho = aqstate.rho/1000;    // the density of water (in g/cm3)
            const auto& epsilon = aqstate.epsilon; // the dielectric constant of water

            // Auxiliary variables
            const auto sqrtI = sqrt(I);
            const auto sqrt_rho = sqrt(rho);
            const auto T_epsilon = T * epsilon;
            const auto sqrt_T_epsilon = sqrt(T_epsilon);
            const auto A = 1.824829238e+6 * sqrt_rho/(T_epsilon*sqrt_T_epsilon);
            const auto B = 50.29158649 * sqrt_rho/sqrt_T_epsilon;
            const auto ln10 = log(10);
            const auto Agamma = 0.5095; // the Debye-Huckel parameter

            // Loop over all ion exchange species in composition
            for(auto i = 0; i < num_species; i++)
            {
                // Fetch phreeqc species from the `attachedData` field of the species
                const auto phreeqc_species = std::any_cast<const PhreeqcSpecies*>(species[i].attachedData());

                // Fetch Debye--Huckel activity model parameters a and b
                const auto a = phreeqc_species->dha;
                const auto b = phreeqc_species->dhb;

                if(b == 99.9) // the Phreeqc hack used while reading the dat-file, which indicate to use Davies activity model
                {
                    // Calculate the ln activity coefficient of the exchange species using the Davies activity model
                    ln_g[i] = ln10*(-Agamma*ze[i]*ze[i]*sqrtI/(1 + sqrtI) - 0.3*I);
                }
                else // otherwise, the Debye--Huckel activity model is preferred
                {
                    // Calculate the ln activity coefficient of the exchange species using the Debye--Huckel model
                    ln_g[i] = ln10*(-A*ze[i]*ze[i]*sqrtI/(1.0 + a*B*sqrtI) + b*I);
                }
            }
        }
        // Add the correction introduced by the activity coefficients
        ln_a += ln_g;

    };
    return fn;
}

} // namespace detail

auto ActivityModelIonExchange() -> ActivityModelGenerator
{
    return ActivityModelIonExchangeGainesThomas();
}

auto ActivityModelIonExchangeGainesThomas() -> ActivityModelGenerator
{
    return [=](const SpeciesList& species)
    {
        return detail::activityModelIonExchangeGainesThomas(species);
    };
}

} // namespace Reaktoro

File: Reaktoro/Thermodynamics/Surface/ActivityModelIonExchange.py
# Reaktoro is a unified framework for modeling chemically reactive systems.
#
# Copyright © 2014-2021 Allan Leal
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this library. If not, see <http://www.gnu.org/licenses/>.


from reaktoro import *
import numpy as np
import pytest


def initializeMoleFractions(species):

    idx = lambda formula: species.indexWithFormula(formula)

    n = 1e-6 * np.ones(species.size())

    n[idx("MgX2")] = 0.1
    n[idx("CaX2")] = 0.2
    n[idx("NaX")]  = 0.3

    return n / n.sum()


def testActivityModelIonExchage():

    db = PhreeqcDatabase("phreeqc.dat")

    # Expected species: AlOHX2 AlX3 BaX2 CaX2 CdX2 CuX2 FeX2 KX LiX MgX2 MnX2 NH4X NaX PbX2 SrX2 ZnX2
    species = db.species().withAggregateState(AggregateState.IonExchange).withCharge(0.0)

    # Check the names of the species
    assert species[0].name()  == "AlOHX2"
    assert species[1].name()  == "AlX3"
    assert species[2].name()  == "BaX2"
    assert species[3].name()  == "CaX2"
    assert species[4].name()  == "CdX2"
    assert species[5].name()  == "CuX2"
    assert species[6].name()  == "FeX2"
    assert species[7].name()  == "KX"
    assert species[8].name()  == "LiX"
    assert species[9].name()  == "MgX2"
    assert species[10].name() == "MnX2"
    assert species[11].name() == "NH4X"
    assert species[12].name() == "NaX"
    assert species[13].name() == "PbX2"
    assert species[14].name() == "SrX2"
    assert species[15].name() == "ZnX2"

    # Initialize input data for the ActivityProps
    T = autodiff.real(300.0)
    P = autodiff.real(12.3e5)
    x = initializeMoleFractions(species)

    # Construct the activity model function with the given ion exchange species.
    fn = ActivityModelIonExchangeGainesThomas()(species)

    # Create the ActivityProps object with the results.
    props = ActivityProps.create(species.size())

    # Evaluate the activity props function
    # fn(props, T, P, x)

    # Fetch exchanger equivalents in the ion exchange species
    #ze = props.extra.at(0)
    #
    # assert ze[0]  == 2
    # assert ze[1]  == 3
    # assert ze[2]  == 2
    # assert ze[3]  == 2
    # assert ze[4]  == 2
    # assert ze[5]  == 2
    # assert ze[6]  == 2
    # assert ze[7]  == 1
    # assert ze[8]  == 1
    # assert ze[9]  == 2
    # assert ze[10] == 2
    # assert ze[11] == 1
    # assert ze[12] == 1
    # assert ze[13] == 2
    # assert ze[14] == 2
    # assert ze[15] == 2

    # props.ln_a[0]  == Approx(-13.017)
    # props.ln_a[1]  == Approx(-12.6116)
    # props.ln_a[2]  == Approx(-13.017)
    # props.ln_a[3]  == Approx(-0.810957)
    # props.ln_a[4]  == Approx(-13.017)
    # props.ln_a[5]  == Approx(-13.017)
    # props.ln_a[6]  == Approx(-13.017)
    # props.ln_a[7]  == Approx(-13.7102)
    # props.ln_a[8]  == Approx(-13.7102)
    # props.ln_a[9]  == Approx(-1.5041)
    # props.ln_a[10] == Approx(-13.017)
    # props.ln_a[11] == Approx(-13.7102)
    # props.ln_a[12] == Approx(-1.09864)
    # props.ln_a[13] == Approx(-13.017)
    # props.ln_a[14] == Approx(-13.017)
    # props.ln_a[15] == Approx(-13.017)
